File: AssetManager/UserInterface/Forms/GK_Updater/ManagePackFile.vb
Imports System.IO
Imports System.Threading

Public Class ManagePackFile
    Public Progress As ProgressCounter
    Public Status As String = ""

    Sub New()
        Progress = New ProgressCounter
    End Sub

    ''' <summary>
    ''' Creates and cleans the pack file directories then downloads a new pack file from the server location.
    ''' </summary>
    ''' <returns></returns>
    Public Async Function DownloadPack() As Task(Of Boolean)
        If Not Directory.Exists(GKPackFileFDir) Then
            Directory.CreateDirectory(GKPackFileFDir)
        End If

        If Directory.Exists(GKExtractDir) Then
            Directory.Delete(GKExtractDir, True)
        End If

        If File.Exists(GKPackFileFullPath) Then
            File.Delete(GKPackFileFullPath)
        End If
        Progress = New ProgressCounter
        Return Await CopyPackFile(GKRemotePackFilePath, GKPackFileFullPath)
    End Function

    ''' <summary>
    ''' Verifies directory structure, checks if pack file is present, then compares local and remote hashes of the pack file.
    '''
    ''' Returns False if directory or file is missing, or if the hashes mismatch.
    ''' </summary>
    ''' <returns></returns>
    Public Async Function VerifyPackFile() As Task(Of Boolean)
        Try
            If Not Directory.Exists(GKPackFileFDir) Then
                Return False
            End If
            If Not Directory.Exists(GKExtractDir) Then
                Return False
            End If
            If Not File.Exists(GKPackFileFullPath) Then
                Return False
            Else

                Dim LocalHash = Await Task.Run(Function()
                                                   Return GetHashOfFile(GKPackFileFullPath)
                                               End Function)

                Dim RemoteHash = Await Task.Run(Function()
                                                    Return GetRemoteHash()
                                                End Function)

                If LocalHash = RemoteHash Then
                    Return True
                Else
                    Return False
                End If
            End If
        Catch ex As Exception
            ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod())
            Return False
        End Try
    End Function

    ''' <summary>
    ''' Returns the contents of the hash text file located in <see cref="GKRemotePackFileDir"/>
    ''' </summary>
    ''' <returns></returns>
    Private Function GetRemoteHash() As String
        Using sr As New StreamReader(GKRemotePackFileDir & GKPackHashName)
            Return sr.ReadToEnd
        End Using
    End Function

    ''' <summary>
    ''' Copies a single file to the <paramref name="dest"/> path.
    ''' </summary>
    ''' <param name="source"></param>
    ''' Path of source file.
    ''' <param name="dest"></param>
    ''' Path of destination.
    ''' <returns></returns>
    Public Async Function CopyPackFile(source As String, dest As String) As Task(Of Boolean)
        If File.Exists(dest) Then
            File.Delete(dest)
        End If
        Return Await Task.Run(Function()
                                  Try
                                      CopyFile(source, dest)
                                      Return True
                                  Catch ex As Exception
                                      Return False
                                  End Try
                              End Function)
    End Function

    ''' <summary>
    ''' Performs a buffered file stream transfer.
    ''' </summary>
    ''' <param name="Source"></param>
    ''' <param name="Dest"></param>
    Private Sub CopyFile(Source As String, Dest As String)
        Dim BufferSize As Integer = 256000
        Dim buffer(BufferSize - 1) As Byte
        Dim bytesIn As Integer = 1
        Dim CurrentFile As New FileInfo(Source)
        Progress.ResetProgress()
        Using fStream As System.IO.FileStream = CurrentFile.OpenRead(),
                destFile As System.IO.FileStream = New FileStream(Dest, FileMode.OpenOrCreate, FileAccess.Write, FileShare.Write, BufferSize, FileOptions.None)
            Progress.BytesToTransfer = CInt(fStream.Length)
            Do Until bytesIn < 1
                bytesIn = fStream.Read(buffer, 0, BufferSize)
                If bytesIn > 0 Then
                    destFile.Write(buffer, 0, bytesIn)
                    Progress.BytesMoved = bytesIn
                End If
            Loop
        End Using
    End Sub

    ''' <summary>
    ''' Compresses the local Gatekeeper directory into a new pack file.
    ''' </summary>
    ''' <returns></returns>
    Public Async Function PackGKDir() As Task(Of Boolean)
        Try
            Progress = New ProgressCounter
            Dim CompDir As New GZipCompress(Progress)
            If Not Directory.Exists(GKPackFileFDir) Then
                Directory.CreateDirectory(GKPackFileFDir)
            End If

            If File.Exists(GKPackFileFullPath) Then
                File.Delete(GKPackFileFullPath)
            End If
            Await Task.Run(Sub()
                               CompDir.CompressDirectory(GKInstallDir, GKPackFileFullPath)
                           End Sub)
            Return True
        Catch ex As Exception
            ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod())
            Return False
        End Try
    End Function

    ''' <summary>
    ''' Decompresses the pack file into a local working directory.
    ''' </summary>
    ''' <returns></returns>
    Private Async Function UnPackGKDir() As Task(Of Boolean)
        Try
            Status = "Unpacking...."
            Progress = New ProgressCounter
            Dim CompDir As New GZipCompress(Progress)
            Await Task.Run(Sub()
                               CompDir.DecompressToDirectory(GKPackFileFullPath, GKExtractDir)
                           End Sub)
            Return True
        Catch ex As Exception
            ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod())
            Return False
        End Try
    End Function

    ''' <summary>
    ''' Copies the pack file and hash file to the server directory.
    ''' </summary>
    ''' <returns></returns>
    Private Async Function UploadPackFiles() As Task(Of Boolean)
        Dim Done As Boolean = False
        Status = "Uploading Pack File..."
        Progress = New ProgressCounter
        Done = Await CopyPackFile(GKPackFileFullPath, GKRemotePackFilePath)

        Status = "Uploading Hash File..."
        Progress = New ProgressCounter
        Done = Await CopyPackFile(GKPackFileFDir & GKPackHashName, GKRemotePackFileDir & GKPackHashName)
        Return Done
    End Function

    ''' <summary>
    ''' Verifies the local pack file and downloads a new one if needed.
    ''' </summary>
    ''' <returns></returns>
    Public Async Function ProcessPackFile() As Task(Of Boolean)
        Dim PackFileOK As Boolean = False
        Status = "Verifying Pack File..."
        If Await VerifyPackFile() Then
            PackFileOK = Await UnPackGKDir()
        Else
            Status = "Downloading Pack File..."
            If Await DownloadPack() Then
                PackFileOK = Await UnPackGKDir()
            End If
        End If
        If PackFileOK Then
            Status = "Done."
            Await Task.Run(Sub()
                               Thread.Sleep(1000)
                           End Sub)
            Return True
        Else
            Status = "ERROR!"
            Return False
        End If

    End Function

    ''' <summary>
    ''' Creates a new pack file and hash file and copies them to the server location.
    ''' </summary>
    ''' <returns></returns>
    Public Async Function CreateNewPackFile() As Task(Of Boolean)
        Try
            Dim OK As Boolean = False
            Status = "Creating Pack File..."
            Progress = New ProgressCounter
            OK = Await PackGKDir()

            Status = "Generating Hash..."
            OK = Await CreateHashFile()

            OK = Await UploadPackFiles()

            If OK Then
                Status = "Done."
            Else
                Status = "Something went wrong..."
            End If
            Return OK
        Catch ex As Exception
            Status = "ERROR!"
            ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod())
            Return False
        End Try
    End Function

    ''' <summary>
    ''' Creates a text file containing the hash string of the pack file.
    ''' </summary>
    ''' <returns></returns>
    Private Async Function CreateHashFile() As Task(Of Boolean)
        If File.Exists(GKPackFileFDir & GKPackHashName) Then
            File.Delete(GKPackFileFDir & GKPackHashName)
        End If
        Dim Hash = Await Task.Run(Function()
                                      Return GetHashOfFile(GKPackFileFullPath)
                                  End Function)
        Using sw As StreamWriter = File.CreateText(GKPackFileFDir & GKPackHashName)
            sw.Write(Hash)
        End Using
        Return True
    End Function

End Class

File: AssetManager/UserInterface/Forms/GK_Updater/ManagePackFile.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.IO;
using System.Threading;

namespace AssetManager.UserInterface.Forms.GK_Updater
{

    public class ManagePackFile
    {
        public ProgressCounter Progress;

        public string Status = "";
        public ManagePackFile()
        {
            Progress = new ProgressCounter();
        }

        /// <summary>
        /// Creates and cleans the pack file directories then downloads a new pack file from the server location.
        /// </summary>
        /// <returns></returns>
        public async Task<bool> DownloadPack()
        {
            if (!Directory.Exists(Paths.GKPackFileFDir))
            {
                Directory.CreateDirectory(Paths.GKPackFileFDir);
            }

            if (Directory.Exists(Paths.GKExtractDir))
            {
                Directory.Delete(Paths.GKExtractDir, true);
            }

            if (File.Exists(Paths.GKPackFileFullPath))
            {
                File.Delete(Paths.GKPackFileFullPath);
            }
            Progress = new ProgressCounter();
            return await CopyPackFile(Paths.GKRemotePackFilePath, Paths.GKPackFileFullPath);
        }

        /// <summary>
        /// Verifies directory structure, checks if pack file is present, then compares local and remote hashes of the pack file.
        ///
        /// Returns False if directory or file is missing, or if the hashes mismatch.
        /// </summary>
        /// <returns></returns>
        public async Task<bool> VerifyPackFile()
        {
            try
            {
                if (!Directory.Exists(Paths.GKPackFileFDir))
                {
                    return false;
                }
                if (!Directory.Exists(Paths.GKExtractDir))
                {
                    return false;
                }
                if (!File.Exists(Paths.GKPackFileFullPath))
                {
                    return false;

                }
                else
                {
                    string LocalHash = await Task.Run(() => { return SecurityTools.GetMD5OfFile(Paths.GKPackFileFullPath); });

                    string RemoteHash = await Task.Run(() => { return GetRemoteHash(); });

                    if (LocalHash == RemoteHash)
                    {
                        return true;
                    }
                    else
                    {
                        return false;
                    }
                }
            }
            catch (Exception ex)
            {
                ErrorHandling.ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod());
                return false;
            }
        }

        /// <summary>
        /// Returns the contents of the hash text file located in <see cref="Paths.GKRemotePackFileDir"/>
        /// </summary>
        /// <returns></returns>
        private string GetRemoteHash()
        {
            using (StreamReader sr = new StreamReader(Paths.GKRemotePackFileDir + Paths.GKPackHashName))
            {
                return sr.ReadToEnd();
            }
        }

        /// <summary>
        /// Copies a single file to the <paramref name="dest"/> path.
        /// </summary>
        /// <param name="source"></param>
        /// Path of source file.
        /// <param name="dest"></param>
        /// Path of destination.
        /// <returns></returns>
        public async Task<bool> CopyPackFile(string source, string dest)
        {
            if (File.Exists(dest))
            {
                File.Delete(dest);
            }
            return await Task.Run(() =>
            {
                try
                {
                    CopyFile(source, dest);
                    return true;
                }
                catch
                {
                    return false;
                }
            });
        }

        /// <summary>
        /// Performs a buffered file stream transfer.
        /// </summary>
        /// <param name="Source"></param>
        /// <param name="Dest"></param>
        private void CopyFile(string Source, string Dest)
        {
            int BufferSize = 256000;
            byte[] buffer = new byte[BufferSize];
            int bytesIn = 1;
            FileInfo CurrentFile = new FileInfo(Source);
            Progress.ResetProgress();
            using (System.IO.FileStream fStream = CurrentFile.OpenRead())
            using (FileStream destFile = new FileStream(Dest, FileMode.OpenOrCreate, FileAccess.Write, FileShare.Write, BufferSize, FileOptions.None))
            {
                Progress.BytesToTransfer = (int)fStream.Length;
                while (!(bytesIn < 1))
                {
                    bytesIn = fStream.Read(buffer, 0, BufferSize);
                    if (bytesIn > 0)
                    {
                        destFile.Write(buffer, 0, bytesIn);
                        Progress.BytesMoved = bytesIn;
                    }
                }
            }
        }

        /// <summary>
        /// Compresses the local Gatekeeper directory into a new pack file.
        /// </summary>
        /// <returns></returns>
        public async Task<bool> PackGKDir()
        {
            try
            {
                Progress = new ProgressCounter();
                GZipCompress CompDir = new GZipCompress(Progress);
                if (!Directory.Exists(Paths.GKPackFileFDir))
                {
                    Directory.CreateDirectory(Paths.GKPackFileFDir);
                }

                if (File.Exists(Paths.GKPackFileFullPath))
                {
                    File.Delete(Paths.GKPackFileFullPath);
                }
                await Task.Run(() => { CompDir.CompressDirectory(Paths.GKInstallDir, Paths.GKPackFileFullPath); });
                return true;
            }
            catch (Exception ex)
            {
                ErrorHandling.ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod());
                return false;
            }
        }

        /// <summary>
        /// Decompresses the pack file into a local working directory.
        /// </summary>
        /// <returns></returns>
        private async Task<bool> UnPackGKDir()
        {
            try
            {
                Status = "Unpacking....";
                Progress = new ProgressCounter();
                GZipCompress CompDir = new GZipCompress(Progress);
                await Task.Run(() => { CompDir.DecompressToDirectory(Paths.GKPackFileFullPath, Paths.GKExtractDir); });
                return true;
            }
            catch (Exception ex)
            {
                ErrorHandling.ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod());
                return false;
            }
        }

        /// <summary>
        /// Copies the pack file and hash file to the server directory.
        /// </summary>
        /// <returns></returns>
        private async Task<bool> UploadPackFiles()
        {
            bool Done = false;
            Status = "Uploading Pack File...";
            Progress = new ProgressCounter();
            Done = await CopyPackFile(Paths.GKPackFileFullPath, Paths.GKRemotePackFilePath);

            Status = "Uploading Hash File...";
            Progress = new ProgressCounter();
            Done = await CopyPackFile(Paths.GKPackFileFDir + Paths.GKPackHashName, Paths.GKRemotePackFileDir + Paths.GKPackHashName);
            return Done;
        }

        /// <summary>
        /// Verifies the local pack file and downloads a new one if needed.
        /// </summary>
        /// <returns></returns>
        public async Task<bool> ProcessPackFile()
        {
            bool PackFileOK = false;
            Status = "Verifying Pack File...";
            if (await VerifyPackFile())
            {
                PackFileOK = await UnPackGKDir();
            }
            else
            {
                Status = "Downloading Pack File...";
                if (await DownloadPack())
                {
                    PackFileOK = await UnPackGKDir();
                }
            }
            if (PackFileOK)
            {
                Status = "Done.";
                await Task.Run(() => { Thread.Sleep(1000); });
                return true;
            }
            else
            {
                Status = "ERROR!";
                return false;
            }

        }

        /// <summary>
        /// Creates a new pack file and hash file and copies them to the server location.
        /// </summary>
        /// <returns></returns>
        public async Task<bool> CreateNewPackFile()
        {
            try
            {
                bool OK = false;
                Status = "Creating Pack File...";
                Progress = new ProgressCounter();
                OK = await PackGKDir();

                Status = "Generating Hash...";
                OK = await CreateHashFile();

                OK = await UploadPackFiles();

                if (OK)
                {
                    Status = "Done.";
                }
                else
                {
                    Status = "Something went wrong...";
                }
                return OK;
            }
            catch (Exception ex)
            {
                Status = "ERROR!";
                ErrorHandling.ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod());
                return false;
            }
        }

        /// <summary>
        /// Creates a text file containing the hash string of the pack file.
        /// </summary>
        /// <returns></returns>
        private async Task<bool> CreateHashFile()
        {
            if (File.Exists(Paths.GKPackFileFDir + Paths.GKPackHashName))
            {
                File.Delete(Paths.GKPackFileFDir + Paths.GKPackHashName);
            }
            object Hash = await Task.Run(() => { return SecurityTools.GetMD5OfFile(Paths.GKPackFileFullPath); });
            using (StreamWriter sw = File.CreateText(Paths.GKPackFileFDir + Paths.GKPackHashName))
            {
                sw.Write(Hash);
            }
            return true;
        }

    }
}
